modules/repository-datastax/src/main/java/com/intuit/wasabi/repository/cassandra/impl/CassandraExperimentRepository.java

```diff
@@ -402,6 +417,28 @@ protected Experiment internalGetExperiment(Application.Name appName,
     }
 
     /**
+     * Updates the tags for a given Application.
+     *
+     * @param applicationName the Application for which the tags should be added
+     * @param tags            list of tags for the experiment
+     */
+    protected void updateExperimentTags(Application.Name applicationName, Set<String> tags) {
+
+        if (tags == null)
+            tags = Collections.EMPTY_SET;
+
+        // update the tag table, just rewrite the complete entry
+        List<ExperimentTagsByApplication> listAllTags = experimentAccessor.getAllTags(applicationName.toString()).all();
```

Comment:
Also @AndreaSuckro , if we really want to support the `delete tag` use case on the application-tag index table, I think a safer and more optimal way of doing that would be to not use a set of tags in one single row but instead have another column for `experiment id`. So, your experiment_tag table would have the `app_name` as the partition key and 2 other columns which are `experiment id` and `tags`. `experiment_id` would be your clustering key. When you update a tag to the experiment, you update the `experiment` table as you are doing right now and you update the `experiment_tag` table in the corresponding row of that `experiment_id`. When you want to retrieve the list of tags for an application, you just query the `experiment_tag` table using the `app_name` as the key and it will return you a result set of all rows containing different experiments and tags corresponding to that `app_name`. You can then merge the list of tags in the code and return. 

That way, you will only be affecting the row for a particular experiment at a time, using atomic operations on the set of tags (either add or delete) and you avoid querying the `experiment` table for updating the `experiment_tag` table.

Does that sound like a reasonable approach?

Apologies for suggesting it too late in the review cycle but I guess it is important to do it the right way. If you do not understand anything, feel free to schedule a BlueJeans session with me if you want.